File: Queue/queue.js
//FIFO - linear data structure

class Queue {
  constructor() {
    this.storage = [];
    this.head = 0;
    this.tail = 0;
  }

  //adding from back - tail is back
  enqueue(element) {
    this.storage[this.tail] = element;
    this.tail++;
    console.log(`added element: ${element}`);
  }

  //method 1: remove element from front/head
  // dequeue() {
  //   let removed = this.storage[this.head];
  //   delete this.storage[this.head];
  //   this.head++;
  //   console.log(`removed element: ${removed}`);
  //   return removed;
  // }

  // method 2: remove element from front/head
  dequeue() {
    if (this.isEmpty() === false) {
      this.tail--;
      console.log(`removed element: ${this.storage.shift()}`);
      return this.storage.shift();
    }
  }

  Length() {
    console.log(`length: ${this.tail}`);
    return this.tail;
  }

  isEmpty() {
    if (this.rear === 0) {
      console.log("The queue is empty!!!");
    }
  }
}

// test queue by performing operations
const queue = new Queue();
queue.enqueue("seahorse");
queue.enqueue("dolphin");
queue.enqueue("whale shark");

queue.dequeue();
queue.dequeue();

queue.Length();
queue.isEmpty();

File: Queue/queue.c
// FIFO - linear data structure

#include <stdio.h>
#define MAX 8

int items[MAX], front = -1, rear = -1;

// added to rear
void enqueue(int value)
{
    if (rear == MAX - 1)
    {
        printf("Queue is full!!!");
    }
    else
    {
        rear++;
        items[rear] = value;
        printf("\n Added element: %d", value);
    }
}

// remove from front
void dequeue(int value)
{
    // if empty, cannot remove
    if (front == rear)
    {
        printf("\n Queue is empty!!!");
    }
    else
    {
        front++;
        value = items[front];
        printf("\n Removed element: %d", value);
    }
}

// test queue by performing operations:
int main()
{
    enqueue(4);
    enqueue(6);
    enqueue(8);

    dequeue(4);
    dequeue(6);
    dequeue(8);

    return 0;
}